src/vcpkg/base/downloads.cpp

```diff
@@ -883,6 +884,110 @@ namespace vcpkg
+
+        // base64 encoded IDs, generated with
+        // for I in @ A B C D E F G H I; do echo '"'$(echo "00${I}" | base64)'",' ; done
+        static const char* block_ids[10] = {
+            "MDBACg==",
+            "MDBBCg==",
+            "MDBCCg==",
+            "MDBDCg==",
+            "MDBECg==",
+            "MDBFCg==",
+            "MDBGCg==",
+            "MDBHCg==",
+            "MDBICg==",
+            "MDBJCg==",
+        };
+
+        if (file_size > block_size * std::extent<decltype(block_ids)>::value) return false;
+
+        auto http_cmd = Command{"curl"}.string_arg("--fail-early");
+
+        auto block_url = raw_url.to_string();
+        block_url += raw_url.contains('?') ? '&' : '?';
+        block_url += "&comp=block&blockid=";
+
+        auto block_file = fmt::format("{}.chunk", file);
+        auto block_id = std::begin(block_ids);
+        auto block_list = std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<BlockList>\n");
+        for (size_t range_from = 0; range_from < file_size; range_from += block_size, ++block_id)
+        {
+            auto range_to = std::min(range_from + block_size, file_size);
+
+            // Extract block data locally
+            http_cmd.string_arg("-o").string_arg(block_file);
+            http_cmd.string_arg("--range").string_arg(fmt::format("{}-{}", range_from, range_to - 1));
+            http_cmd.string_arg(fmt::format("file://{}", file));
+            http_cmd.string_arg("--next");
+
+            // Upload block
+            add_curl_headers(http_cmd, headers);
+            http_cmd.string_arg("-X").string_arg("PUT");
+            http_cmd.string_arg(block_url + *block_id);
+            http_cmd.string_arg("-T").string_arg(block_file);
+            http_cmd.string_arg("--next");
+            http_cmd.string_arg("-w").string_arg("\\n" + guid_marker.to_string() + "%{http_code}");
+
+            block_list += "  <Latest>";
+            block_list += *block_id;
+            block_list += "</Latest>\n";
+        }
+        block_list += "</BlockList>\n";
+
+        std::error_code ec;
+        auto block_list_filepath = file + ".blocklist";
+        fs.write_contents(block_list_filepath, block_list, ec);
+        if (ec) return false;
+
+        auto block_list_url = raw_url.to_string();
+        block_list_url += raw_url.contains('?') ? '&' : '?';
+        block_list_url += "comp=blocklist";
+        add_curl_headers(http_cmd, {}); // do not send x-ms-blob-type!
+        http_cmd.string_arg("-H").string_arg("x-ms-version: 2020-04-08");
+        http_cmd.string_arg("-X").string_arg("PUT");
+        http_cmd.string_arg(block_list_url);
+        http_cmd.string_arg("-T").string_arg(block_list_filepath);
+        http_cmd.string_arg("-w").string_arg("\\n" + guid_marker.to_string() + "%{http_code}");
+
+        int code = 0;
+        auto res = cmd_execute_and_stream_lines(context, http_cmd, [&code](StringView line) {
```

Comment:
Is there a concern about parallel access to the blob storage? How does this handle cases where multiple vcpkg instances access the blob storage and attempt to upload blocks with the same IDs but for different blobs (for example, both try to upload a block with ID `MDBACg==`)? Should we use the package SHA to ensure sufficiently unique IDs?

Maybe we need a final integrity check once all blocks have been uploaded (In which case, maybe the SHA of the complete blob can be used to ensure uniqueness).

Reply:
IMO (the same version of) vcpkg would create the same blocks for any artifact+block ID. 
So successful commits will always find the same block content as long as Azure doesn't pick incomplete uploads.

But right now I am testing an implementation with `azcopy`.

Reply:
> But right now I am testing an implementation with `azcopy`.

#1679

Reply:
For clarity: Block IDs are scoped per blob (vcpkg: ABI hash).  
Concurrency exists only with regard to parallel uploads of the same ABI hash.  
And we talking about large artifacts which take a while for an upload to complete.